llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -16749,20 +16745,36 @@ static SDValue performSETCCCombine(SDNode *N,
           combineVectorSizedSetCCEquality(VT, N0, N1, Cond, dl, DAG, Subtarget))
     return V;
 
-  // (X & -4096) == 0 -> (X >> 12) == 0 if the AND constant can't use ANDI.
   if (DCI.isAfterLegalizeDAG() && isNullConstant(N1) &&
       N0.getOpcode() == ISD::AND && N0.hasOneUse() &&
       isa<ConstantSDNode>(N0.getOperand(1))) {
     const APInt &AndRHSC =
         cast<ConstantSDNode>(N0.getOperand(1))->getAPIntValue();
+    // (X & -4096) == 0 -> (X >> 12) == 0 if the AND constant can't use ANDI.
```

Comment:
This comment needs improved.  I know you didn't add this, but still.

Something like, handle and with 111..000 mask via (X >> C) == 0?

Reply:
I did add it in a previous patch. It was modified from this comment from the generic transform https://github.com/llvm/llvm-project/blob/a34b110797e0793dd29c68ccad4927c10b7e6c70/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L5365